### PR TITLE
Soar 9298 Microsoft Defender Hunting Plugin

### PR DESCRIPTION
**Pull request by llaszuk-r7**
### Description

Describe the proposed changes:
 - New plugin
 - Added actions: Advanced Hunting (Query) 

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [X] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [X] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [X] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [X] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [X] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [X] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [X] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [X] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [X] Work fully completed
- [X] Functional
  - [X] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [X] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [X] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [X] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [X] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### Run

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "IconValidator" failed! 
        Cause: icon.png file not included in plugin.
```

----
[*] Total time elapsed: 11665.146ms

<summary>
icon-validate --all .
</summary>

</details>

### CODE REVIEW

**Review by igorski-r7 — APPROVED** [review]
Looks good! 😃

**Review by cmcnally-r7 — APPROVED** [review]

File: plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/util/api.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        logger: Optional[Logger] = None,
+    ):
+        super().__init__()
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.tenant_id = tenant_id
+        self.logger = logger
+
+    def advanced_hunting(self, query: str) -> dict:
+        self.connection()
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/util/api.py:91]: You could move it to a decorator like in Igors PR :)

File: plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/util/tools.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                        )
+                        retry = True
+            return func(self, *args, **kwargs)
+
+        return _wrapper
+
+    return _decorate
+
+
+def return_non_empty(input_dict: Dict[str, Any]) -> Union[Dict[Any, Any], Any]:
```
Comment (jkaluza-rapid7) [inline plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/util/tools.py:73]: I truly regret naming this function like that xd it should have been baptised `clean_dictionary_recursively`

**Review by jkaluza-rapid7 — CHANGES_REQUESTED** [review]

File: plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/connection/connection.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    def __init__(self):
+        super(self.__class__, self).__init__(input=ConnectionSchema())
+
+    def connect(self, params):
+        self.logger.info("Connect: Connecting...")
+        self.client = MicrosoftDefenderClientAPI(
+            params.get(Input.CLIENT_ID),
+            params.get(Input.CLIENT_SECRET).get("secretKey"),
+            params.get(Input.TENANT_ID),
+            self.logger,
```
Comment (blaxminarayan-r7) [inline plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/connection/connection.py:18]:
```suggestion
            self.logger
```
```suggestion
            self.logger,
```

Comment (blaxminarayan-r7) [inline plugins/microsoft_defender_hunting/icon_microsoft_defender_hunting/connection/connection.py:18]:
```suggestion
            self.logger
```
```suggestion
            self.logger,
```

**Pull request merged**